### ISSUE TITLE
RSS/HTML import doesn't handle newlines

### ISSUE THREAD

**Issue by binford2k** [thepracticaldev/dev.to#8457]
**Describe the bug**

HTML does not distinguish whitespace. This means that text like so

```
This is a line
with some
newline characters.
```

and this

`This is a line with some newline characters.`

should be treated exactly the same. Unfortunately, the RSS importer just strips the newline character, so the resulting string becomes this:

`This is a linewith somenewline characters.`

This makes syndicated content a GIANT pain to post.

In my case, I publish via Jekyll and GitHub pages and I also wrap my markdown source. See an example at https://raw.githubusercontent.com/binford2k/binford2k.com/master/_posts/2020-06-10-writing-a-new-metrics-plugin.md

Jekyll publishes the text with embedded newlines, which is rendered properly by every browser that correctly implements the HTML spec, including the browser in which you're reading this issue.

``` html
<p>I’ve been working on the Dropsonde telemetry framework for the Puppet ecosystem
for a while. If you’ve followed any of its development, you likely already know
that the main focus is on providing community value and maintaining privacy and
transparency.</p>
``` 

ends up imported as

> I’ve been working on the Dropsonde telemetry framework for the Puppet ecosystemfor a while. If you’ve followed any of its development, you likely already knowthat the main focus is on providing community value and maintaining privacy andtransparency.

**To Reproduce**

1. Publish content with newline wrapped text.
1. RSS import it.

**Expected behavior**

Newline characters should be treated like whitespace, following the HTML spec.

**Comment by rhymes** [comment]
Thanks for the issue! We'll take your request into consideration and follow up if we decide to tackle this issue.

To our amazing contributors: issues labeled `type: bug` are always up for grabs, but for feature requests, please wait until we add a `ready for dev` before starting to work on it.

To claim an issue to work on, please leave a comment. If you've claimed the issue and need help, please ping @thepracticaldev/oss and we will follow up within 3 business days.

For full info on how to contribute, please check out our [contributors guide](https://github.com/thepracticaldev/dev.to/blob/master/CONTRIBUTING.md).

**Comment by rhymes** [comment]
@binford2k would the following comment - https://github.com/thepracticaldev/dev.to/issues/7923#issuecomment-630399659 - help?

AFAIK we don't support markdown in the RSS feed, the content should be HTML

**Comment by citizen428** [comment]
@rhymes The feed doesn't seem to include Markdown, but HTML:

http://binford2k.com/atom.xml

**Comment by binford2k** [comment]
Honestly, it looks like `\n` characters are stripped out, and if they were instead just replaced with a `[space]`, it'd be all good.

**Comment by michael-tharrington** [comment]
Another DEV user just recently wrote in to confirm that they are experiencing this same problem. They linked to this article written by a separate user - https://dev.to/jhermann/bug-in-article-conversion-from-rss-feeds-3gae — which seems to confirm that there are at least 3 folks experiencing this. Just wanted bring this up so we can get a handle on how often this occurs + if these different RSSes have anything in common:
- jhermann.github.io/blog/feed.xml
- hugopeixoto.net/articles.xml

**Comment by hugopeixoto** [comment]
:wave: I'm the reporter Michael mentioned.

I'm trying to reproduce this locally. This is how the imported markdown for [this post](https://hugopeixoto.net/articles/rust-gamedev-ecs-bevy.html) looks like:

```
One week later, [SGDQ - Summer Games Done Quick](https://gamesdonequick.com/)happened - a bi-yearly speedrunning marathon that lasts for a week, raising funds for[Médecins Sans Frontières](https://en.wikipedia.org/wiki/M%C3%A9decins_Sans_Fronti%C3%A8res).To avoid spending 168 hours watching a twitch stream, [eevee](https://eev.ee/) started a game jam:[Games made quick](https://itch.io/jam/games-made-quick-four-plus).

I’ve tried to make some games before, but I never released anything. I strugglea lot with the code architecture and end up obsessing about it instead oftrying to create a game.
```

There are two problems here. In the first paragraph, newlines after or before links are removed. In the second paragraph, you see that newlines between lines are also removed (`strugglea`, `oftrying`).

I was able to reproduce this, but I don't grok everything yet. Here's a `bin/rails c` session:

```ruby
[1] pry(main)> require_relative './app/lib/reverse_markdown/converters/custom_text.rb'
=> true
[2] pry(main)> ReverseMarkdown::Converters.lookup(:html).convert(Nokogiri::HTML(
               "a\n<a href='1'>link</a>\nis good\nbut blankspace is better").root)
=> "\n\na[link](1)is goodbut blankspace is better\n\n"
```

If I don't require the `custom_text` converter, I get a different result:

```ruby
[1] pry(main)> ReverseMarkdown::Converters.lookup(:html).convert(Nokogiri::HTML(
               "a\n<a href='1'>link</a>\nis good\nbut blankspace is better").root)
=> "\n\na[link](1)is good but blankspace is better\n\n"
```

So it seems that the blankspace surrounding tags is a problem in the original `ReverseMarkdown` converters. I'll see if I can make a PR to `ReverseMarkdown` to fix the link issue.

The second issue, newlines being all stripped, is a problem of the `custom_text` converter.

The `custom_text` converter seems to be a monkey patched version from 2018. The commit message has an issue reference, but it doesn't seem to be from github:

> Monkey-patched ReverseMarkdown's :text & :pre (#353)

It seems that the change in the monkey patch was what broke this. `remove_inner_newlines` was changed from replacing newlines with spaces to replacing newlines with nothing.

Does anyone have access to the right #353 so that we can see what was the problem that the monkey patch was trying to fix?

**Comment by maestromac** [comment]
Thanks for the thorough dive @hugopeixoto ! You are right, `CustomText` is the source of this issue.  

> Does anyone have access to the right #353 so that we can see what was the problem that the monkey patch was trying to fix?

I do but there's not a lot of information on why. That's my fault; I was new and naive 😅. What I think I was doing is removing the `#preserve_tags` to resolve a code block issue. The `<` and `>` were being escaped within code block. 

> PS: I was having trouble reproducing this locally, probably because the custom_text converter doesn't get autoloaded in development mode, but does get loaded in production mode. IMO this is something that should also be addressed.

I think what we need to do here is to force register with `ReverseMarkdown::Converters.register :Text, ReverseMarkdown::Converters::CustomText.new` at the bottom of those files. That should solve it for development.

The solution would be to [apply the same change](https://github.com/xijo/reverse_markdown/commit/4c1b8f5ffdfe368aa2f6b9b7de74cd2cae78b80d#diff-fc717bcdd23a9f1459a4e00ce90b5407R51) to our `CustomText`. Would you like to a submit a PR? if not, I can go ahead and do it.

**Comment by hugopeixoto** [comment]
> I think what we need to do here is to force register with `ReverseMarkdown::Converters.register :Text, ReverseMarkdown::Converters::CustomText.new` at the bottom of those files. That should solve it for development.

These files already have `register :text, CustomText.new` at the bottom. Since the problem is that they're not being required, changing the file itself won't make a difference. I think we need to add the `require`s to an initializer.

I'll make a PR adding a `config/initializers/reverse_markdown.rb` initializer requiring those two files. What do you think?

> The solution would be to [apply the same change](https://github.com/xijo/reverse_markdown/commit/4c1b8f5ffdfe368aa2f6b9b7de74cd2cae78b80d#diff-fc717bcdd23a9f1459a4e00ce90b5407R51) to our `CustomText`. Would you like to a submit a PR? if not, I can go ahead and do it.

Yeah, I can make a PR fixing `remove_inner_newlines`.

> What I think I was doing is removing the `#preserve_tags` to resolve a code block issue. The `<` and `>` were being escaped within code block.

The `<`/`>` thing was added in https://github.com/xijo/reverse_markdown/pull/58, so by removing `preserve_tags` we may be reintroducing another issue. If the idea was to fix something inside code blocks, maybe the `code` converter should have been overriden instead? Not really sure how this works. I'll see if I can make some test cases for this.

Even if the custom_text issue is fixed, we'll be left with the upstream issue with links being smooshed with surrounding text. I opened an issue (https://github.com/xijo/reverse_markdown/issues/91), but I don't see how to easily solve the problem.

**Comment by mikeydorje** [comment]
Closing as resolved. PR #10476 merged in September 2020 fixing RSS import newline handling in the CustomText converter.

If RSS imports are still removing newlines incorrectly, please open a new issue with reproduction details.
